cli/azd/pkg/experimentation/assignment.go

```diff
@@ -0,0 +1,186 @@
+// AssignmentManager caches the assignment information for 24 hours.
+
+func NewAssignmentsManager(endpoint string, transport policy.Transporter) (*AssignmentsManager, error) {
+	configRoot, err := config.GetUserConfigDir()
+	if err != nil {
+		return nil, err
+	}
+
+	cacheRoot := filepath.Join(configRoot, "experiments")
+	if err := os.MkdirAll(cacheRoot, osutil.PermissionDirectory); err != nil {
+		return nil, err
+	}
+
+	client, err := newTasClient(endpoint, &policy.ClientOptions{
+		Transport: transport,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &AssignmentsManager{
+		cacheRoot: cacheRoot,
+		client:    client,
+	}, nil
+}
+
+// Assignment is a subset of the information returned by the TAS service.
+type Assignment struct {
+	Features          []string
+	Flights           map[string]string
+	Configs           []AssignmentConfig
+	ParameterGroups   []string
+	AssignmentContext string
+}
+
+// AssignmentConfig is information about a specific config in an assignment.
+type AssignmentConfig struct {
+	ID         string
+	Parameters map[string]interface{}
+}
+
+// Assignment gets a the assignment information for this given machine.
+func (am *AssignmentsManager) Assignment(ctx context.Context) (*Assignment, error) {
+	machineId := resource.MachineId()
+
+	cachedAssignment, err := am.readResponseFromCache(machineId)
+	if err != nil {
+		log.Printf("could not read assignment from cache: %v", err)
+	}
+
+	if cachedAssignment == nil {
+		assignment, err := am.client.GetVariantAssignments(ctx, &variantAssignmentRequest{
+			Parameters: map[string]string{
+				MachineIdParameterName: resource.MachineId(),
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		if err := am.cacheResponse(machineId, assignment); err != nil {
+			log.Printf("failed to cache assignment response: %v", err)
+		}
+
+		cachedAssignment = assignment
+	}
+
+	var configs []AssignmentConfig
+
+	if cachedAssignment.Configs != nil {
+		configs = make([]AssignmentConfig, len(cachedAssignment.Configs))
+
+		for i, config := range cachedAssignment.Configs {
+			configs[i] = AssignmentConfig{
+				ID:         config.ID,
+				Parameters: config.Parameters,
+			}
+		}
+	}
+
+	return &Assignment{
+		Features:          cachedAssignment.Features,
+		Flights:           cachedAssignment.Flights,
+		Configs:           configs,
+		ParameterGroups:   cachedAssignment.ParameterGroups,
+		AssignmentContext: cachedAssignment.AssignmentContext,
+	}, nil
+}
+
+// errCacheExpired is returned when the cached data is out of date.
+var errCacheExpired = errors.New("cache expired")
+
+// errUnsupportedCacheVersion is returned with the cache version is not supported
+var errUnsupportedCacheVersion = errors.New("unsupported cache version")
+
+// cacheResponse caches the response from the TAS service for 24 hours, using the machineId as the cache key.
+func (am *AssignmentsManager) cacheResponse(machineId string, response *treatmentAssignmentResponse) error {
```

Comment:
Curious on the motivation to use `machineId` as the cache key. The config directory is user-scoped and not machine-scoped by default, so we're not making the file more unique than it already is. In that sense, `machineId` might give a false sense that this thing is tied to the machine, but in reality it's user-scoped.

Reply:
This is a good question, I should have called out my thinking here.

This was motivated by the behavior I observed with the original implementation of the tas client, from the ExP team.  In their design, the cache key was the entire request object (which makes sense, since all of these parameters could be used by the assignment service to compute your assignment context).

In their implementation of the cache, they used a high performance in memory map, and the key was a string based on hashing all the members of the request.

Here, we only set a single member on the request and so I figured it was "natural" to use that as the key for the cache (since it uniquely identifies the request).  However, given that we want to pass some additional parameters (i.e. the CLI version), we need a better strategy.

What do you think about just changing the cache format to be an object with two properties: "version" and "entries".  entries is an array of objects, each with a "request", and "response" property, as well as "expiresOn".  We expire stale entries on save of the file.  We can compare the request objects in the natural way (two keys are equal if all their properties are the same, ignoring ordering in the arrays and maps because they are not semantically ordered)

That would allow us to use additional stuff in the assignment request later (perhaps we'll want to expose the ability to control the blocked and must be assigned flights via `azd config` at some point?), without having to go back and revisit the cache.  Thoughts?

Reply:
I see. The way I read the motivation behind the ExP client library is to support multitenancy, since there could be more than one user with different user-contexts. I'm not sure if we do run afoul of anything if we kept a simple `assignment.cache` that is refreshed periodically.

Reply:
> I'm not sure if we do run afoul of anything if we kept a simple assignment.cache that is refreshed periodically.

SGTM.